### ISSUE TITLE
Allow users to upload a folder

### ISSUE THREAD

**Issue by albertmichaelj** [jupyterlab/jupyterlab#6037]
Currently, users can drag and drop files onto the sidebar to upload them, but they cannot upload folders. Nor can they upload folders through the upload dialogue menu. I think it would be a significant enhancement if users could upload folders through the drag and drop interface.

I apologize if this is an open issue, but I was unable to find it.

**Comment by ellisonbg** [comment]
Yes!

**Comment by Madhu94** [comment]
I'm taking this up !

**Comment by albertmichaelj** [comment]
@Madhu94 You're my hero! This is literally the biggest need I have for making JupyterLab the perfect education environment for my students. Let me know if you need a tester. Also, if you point me to the right part of the code base, I'm happy to hack away at some parts of it (I honestly couldn't figure out how to start).

Thanks!

**Comment by Madhu94** [comment]
I don't think this is possible without resorting to non standard browser APIs - like [this one](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/webkitdirectory).
Does anyone know of a better solution?

**Comment by albertmichaelj** [comment]
Huh, that's disappointing. I do know it's quite common for web apps to let
you upload folders (at least on chrome), so the API seems to be relatively
common. Does this gracefully fail on non supported browsers?

Given how useful folder upload would be, I would encourage a non-standard
implementation to no implementation. Alternatively, could this be
(relatively easily) implemented as an extension? That way this non standard
API wouldn't be used by the core Jupyter lab code.

**Comment by albertmichaelj** [comment]
@Madhu94 I just wanted to see if you had any further thoughts on this. Based on the reading I've done, it seems like it is fairly common practice to use these non standard browser APIs for folder uploading, and that the one you link to fails gracefully (i.e. if the browser doesn't support it you just get identical behavior to what we currently have). I think a non standard browser API that fails gracefully is better than the current behavior. I feel this way because I strongly feel that this improvement is really meaningful for the user experience of JupyterLab (especially when it is used as part of a jupyterhub).

Are you willing to try to implement it? If not, could you point me to the relevant part of the code, and I may take a stab at it.

Thanks for being willing to take this on!

**Comment by Madhu94** [comment]
@albertmichaelj I'm sorry I haven't attempted this after my first shot, was hoping a maintainer would chime in. If you want to take a shot at this, you can look at the filebrowser extension - specifically [here](https://github.com/jupyterlab/jupyterlab/blob/79a332d517af21c2112d2a830e3b17792fe20e1a/packages/filebrowser/src/listing.ts#L1019). An ideal solution would be to implement this as a lab extension, so that you don't have to worry if it gets accepted into lab's core or not. But I can't see how you could do that for this case without duplicating a lot of what filebrowser does.

**Comment by rufytn** [comment]
@albertmichaelj did you find a solution?

**Comment by krassowski** [comment]
#10596 solves this issue although it relies on experimental API.

**Comment by albertmichaelj** [comment]
@rufytn I never found a direct solution, and I never did work on modifying the code. It does look like #10596 would solve the problem when/if it gets accepted. However, my work around has been to use the [jupyter archive](https://github.com/jupyterlab-contrib/jupyter-archive) extension. This allows me to upload zipped folders and extract them in Jupyter Lab. This solves the problem well enough for me for now.

**Comment by davidbrochart** [comment]
It seems that a solution would be to implement a standard API like WebDAV in jupyter-server.
Jupyverse has an [open PR](https://github.com/jupyter-server/jupyverse/pull/192) for that.

**Comment by krassowski** [comment]
Closing as solved in 4.3 by #10596.
